### PR TITLE
Assignment1 irfanhazarie

### PR DESCRIPTION
**Pull request by irfanhzre**
### **The addressed issue**
This pull request is mainly about optimizing codebase for a better performance. Basically for what I have done, I have added some exception flow if main flow is going wrong using tryCatch function. This mainly to avoid what i have changed become worse.

### **What I have reengineered**

1. CSS and Resource Management

On Previous code, there are two lines of code that generated two different output CSS files. 
```
# Applies css to rShiny app
sass(
  sass_file("www/stylesheets/app.scss"),
  output = "www/stylesheets/app.css"
)

# Generates css used on the about page
# 

> See app.scss file for manual steps needed here

sass(
  sass_file("www/stylesheets/app.scss"),
  output = "about/app.css"
)
```

New Code
Instead of generate two different css files, we can optimize it by generate 1 css and use it in 'about' and 'shinyApp'.
```
sass(sass_file("www/stylesheets/app.scss"), output = "www/stylesheets/app.css")
file.copy("www/stylesheets/app.css", "about/app.css")
```
**Impact of Changes**
By generating only one CSS file, the code is cleaner and avoids redundant operations. Plus, it also reduces the need for multiple sass function calls and file writes in application

2. Error Handling and Data Cleaning Optimization

Old Code:
```
m_reg <- as.data.frame(
  read_excel(file.path(data_folder, "Metadata.xlsx"), sheet = 1)
)
```
New Code:
```
tryCatch({
  m_reg <- as.data.frame(read_excel(file.path(data_folder, "Metadata.xlsx"), sheet = 1))
  m_reg <- m_reg[!is.na(m_reg$Numerator),]
  rownames(m_reg) <- m_reg$Numerator
}, error = function(e) {
  cat("Error loading measure registry: ", e$message, "\n")
  stop("Failed to load measure registry from 'Metadata.xlsx'. Ensure the file exists and is properly formatted.")
})
```
The new code wraps the data loading process in a tryCatch block to handle any errors that might occur during the reading of the Metadata.xlsx file. This helps prevent the application from crashing and provides a descriptive error message if the file cannot be loaded. After successfully loading the data, rows with missing Numerator values are filtered out (m_reg <- m_reg[!is.na(m_reg$Numerator),]). The Numerator column is used to set the row names (rownames(m_reg) <- m_reg$Numerator), ensuring the data is properly indexed.

**Impact of Changes**
The error handling ensures that if there are any issues with the Metadata.xlsx file such as missing file or incorrect format, a clear error message is shown and the program doesn't fail silently. By removing rows with missing values in the Numerator column and setting appropriate row names, the code will ensure that the dataset is clean and well-structured for subsequent operations.

3. Faster file reading

The code for loading data has been optimized by using data.table's fread() function for faster CSV reading, along with error handling to ensure robustness.

Old code:
```
# load index weighting/output
info_df <- read.csv(
  file.path(data_folder, "Cleaned", "HSE_MWI_Data_Information.csv"),
)
rownames(info_df) <- info_df$Numerator

# load mapped measure data excat values
meas_df <- read.csv(
  file.path(data_folder, "Cleaned", "HSE_MWI_ZCTA_Converted_Measures.csv"),
  colClasses = c("GEOID" = "character")
)
meas_df <- meas_df[meas_df$GEOID != "",]
rownames(meas_df) <- meas_df$GEOID
```
New code:
```
# Use data.table for faster CSV reading
library(data.table)

# Load index weighting/output with error handling
tryCatch({
  info_df <- fread(
    file.path(data_folder, "Cleaned", "HSE_MWI_Data_Information.csv"),
    colClasses = list(character = "Numerator")
  )
  setkey(info_df, Numerator) # More efficient than rownames
}, error = function(e) {
  cat("Error loading info_df: ", e$message, "\n")
  # Fallback to read.csv if fread fails
  info_df <- read.csv(
    file.path(data_folder, "Cleaned", "HSE_MWI_Data_Information.csv")
  )
  rownames(info_df) <- info_df$Numerator
  return(info_df)
})

# Load mapped measure data with error handling
tryCatch({
  meas_df <- fread(
    file.path(data_folder, "Cleaned", "HSE_MWI_ZCTA_Converted_Measures.csv"),
    colClasses = c("GEOID" = "character")
  )
  # Filter and set key in one step
  meas_df <- meas_df[GEOID != ""]
  setkey(meas_df, GEOID)
}, error = function(e) {
  cat("Error loading meas_df: ", e$message, "\n")
  # Fallback to read.csv if fread fails
  meas_df <- read.csv(
    file.path(data_folder, "Cleaned", "HSE_MWI_ZCTA_Converted_Measures.csv"),
    colClasses = c("GEOID" = "character")
  )
  meas_df <- meas_df[meas_df$GEOID != "",]
  rownames(meas_df) <- meas_df$GEOID
  return(meas_df)
})
```
The fread() function from the data.table package has been used for faster CSV reading, replacing the slower read.csv(). This is especially beneficial for large datasets.
tryCatch() blocks also have been introduced to handle any errors during the loading of CSV files. If an error occurs such file not found, read error, a descriptive error message is printed, and the program attempts to load the data using the fallback method (read.csv()).

setkey() is used to set the key for both info_df and meas_df to the relevant columns (Numerator and GEOID), which is more efficient than setting rownames(). This improves subsequent data manipulation and joins.
```
setkey(info_df, Numerator)
setkey(meas_df, GEOID)
```
**Impact of Changes**
The use of fread() improves performance when reading large CSV files, reducing the time it takes to load data. tryCatch() ensures that if fread() fails for any reason, the code gracefully falls back to using read.csv(), preventing the application from crashing and providing meaningful error messages. 

The new code ensures better performance, robustness, and efficiency in handling large datasets, with appropriate fallbacks and meaningful error messages when issues arise.

**Reengineering strategy and approach**
Iterative Approach
The incremental approach focuses on making small, iterative improvements to existing processes or systems rather than a full, radical redesign all at once. It involves making gradual changes and continuously improving based on feedback and testing. These changes improve the existing codebase without requiring a full, disruptive redesign. Incremental changes build on the existing framework rather than replacing it entirely.

In summary, this changes reflect the incremental approach in software development and maintainance. By introducing small, manageable enhancements like using fread(), adding error handling, and using setkey()), it optimized the system for better performance, reliability, and efficiency without undertaking a large-scale redesign. This mirrors the incremental improvements often seen in business process reengineering, where the focus is on making continuous, manageable improvements over time.

### CODE REVIEW

**Comment by suhadaudd11** [comment]
Okay
